share/scripts/sdcdb.tcl

```diff
@@ -0,0 +1,795 @@
+}
+
+proc break_fline {filename linenum cond cmd} {
+    set record [search_file $filename $linenum]
+    if {$record eq {}} {
+        error "address not found"
+    } else {
+        debug breakpoint create -address [lindex $record 1] -condition $cond -command $cmd
+    }
+}
+
+proc break_ffunc {filename funcname cond cmd} {
+    lassign [search_func $filename $funcname] start {}
+    if {$start eq {}} {
+        error "function not found"
+    } else {
+        debug breakpoint create -address $start -condition $cond -command $cmd
+    }
+}
+
+proc break_func {funcname cond cmd} {
+    lassign [search_func {} $funcname] start {}
+    if {$start eq {}} {
+        error "function not found"
+    }
+    debug breakpoint create -address $start -condition $cond -command $cmd
+}
+
+proc check {} {
+    variable old_PC
+    return [reg PC] != $old_PC
+}
+
+proc print_status {file line} {
+    puts "file: $file:$line, position: 0x[h [reg PC]]"
+    list_pc -1 1
+}
+
+proc update_step {{type {step}}} {
+    set record [find_source [reg PC]]
+    variable current_file
+    variable current_line
+    if {$record ne {} && $record ne [list $current_file $current_line]} {
+        lassign $record current_file current_line
+        variable times_left
+        incr times_left -1
+        if {$times_left <= 0} { print_status $current_file $current_line }
+        debug break
+    }
+    variable times_left
+    if {$times_left > 0} {
+        # still looping
+        after break "sdcdb::update_step $type"
+        $type  ;# call "step" or "step_over"
+    }
+}
+
+proc find_source {address} {
+    variable addr2file
+    # get current source file
+    set record [lindex [array get addr2file [reg PC]] 1]
+    return $record
+}
+
+proc prepare_step {{n 1}} {
+    debug break  ;# just to be sure
+    # get current position in source code
+    set record [find_source [reg PC]]
+    lassign $record file line
+    lassign [search_file $file $line] begin end
+    debug_out "search_file: $file, $line -> [search_file $file $line]"
+    if {$begin ne {} && $end ne {}} {
+        variable current_file $file
+        variable current_line $line
+    } else {
+        puts "Possibly out of scope, skipping till we get back."
+    }
+    variable times_left $n
+}
+
+proc sdcdb_step {{n 1}} {
+    after break "sdcdb::update_step"
+    prepare_step $n
+    debug step
+}
+
+proc sdcdb_next {{n 1}} {
+    after break "sdcdb::update_step step_over"
+    prepare_step $n
+    step_over
+}
+
+proc print_info {} {
+    variable addr2file
+    if {[array get addr2file [reg PC]] eq {}} {
+        error "line mapping not found for 0x[h [reg PC]]"
+    }
+    lassign $addr2file([reg PC]) file line
+    if {$file eq {}} {
+        error "address [h [reg PC]] not found in database."
+    }
+    print_status $file $line
+}
+
+proc sdcdb_info {arg} {
+    if {$arg eq "-break"} {
+        debug break
+    }
+    if {[catch {print_info} fid]} {
+        puts $fid
```

Comment:
And also here?
BTW this only converts an error message into a non-error return string, is that right?